File: test-projects/java-sample/src/main/java/com/codegraph/sample/UserService.java
package com.codegraph.sample;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Service for managing user operations
 */
public class UserService {
    private final DatabaseConnection db;
    private final Logger logger;

    /**
     * Create a new UserService
     *
     * @param db Database connection instance
     * @param logger Logger instance
     */
    public UserService(DatabaseConnection db, Logger logger) {
        this.db = db;
        this.logger = logger;
    }

    /**
     * Create a new user
     *
     * @param email User's email address
     * @param name User's full name
     * @return The created user
     * @throws IllegalArgumentException If email or name is invalid
     */
    public User createUser(String email, String name) {
        logger.info("Creating user: " + email);

        if (email == null || email.isEmpty() || !email.contains("@")) {
            throw new IllegalArgumentException("Invalid email address");
        }

        if (name == null || name.length() < 2) {
            throw new IllegalArgumentException("Invalid name");
        }

        User user = new User(generateId(), email, name);

        db.execute(
            "INSERT INTO users (id, email, name, created_at) VALUES (?, ?, ?, NOW())",
            user.getId(), user.getEmail(), user.getName()
        );

        logger.info("User created: " + user.getId());
        return user;
    }

    /**
     * Get user by ID
     *
     * @param userId User ID
     * @return User if found, empty otherwise
     */
    public Optional<User> getUserById(String userId) {
        logger.debug("Fetching user: " + userId);

        List<User> results = db.query(
            "SELECT * FROM users WHERE id = ?",
            userId
        );

        return results.isEmpty() ? Optional.empty() : Optional.of(results.get(0));
    }

    /**
     * Update user information
     *
     * @param userId User ID
     * @param email New email address (optional)
     * @param name New name (optional)
     * @return Updated user
     * @throws IllegalStateException If user not found
     */
    public User updateUser(String userId, String email, String name) {
        logger.info("Updating user: " + userId);

        List<String> fields = new ArrayList<>();
        List<Object> values = new ArrayList<>();

        if (email != null) {
            fields.add("email = ?");
            values.add(email);
        }

        if (name != null) {
            fields.add("name = ?");
            values.add(name);
        }

        if (fields.isEmpty()) {
            throw new IllegalArgumentException("No fields to update");
        }

        values.add(userId);

        db.execute(
            "UPDATE users SET " + String.join(", ", fields) + " WHERE id = ?",
            values.toArray()
        );

        return getUserById(userId)
            .orElseThrow(() -> new IllegalStateException("User not found: " + userId));
    }

    /**
     * Delete a user
     *
     * @param userId User ID to delete
     */
    public void deleteUser(String userId) {
        logger.info("Deleting user: " + userId);
        db.execute("DELETE FROM users WHERE id = ?", userId);
    }

    /**
     * List all users with pagination
     *
     * @param limit Maximum number of users to return
     * @param offset Number of users to skip
     * @return List of users
     */
    public List<User> listUsers(int limit, int offset) {
        logger.debug("Listing users: limit=" + limit + ", offset=" + offset);

        return db.query(
            "SELECT * FROM users ORDER BY created_at DESC LIMIT ? OFFSET ?",
            limit, offset
        );
    }

    /**
     * Search users by email or name
     *
     * @param query Search query
     * @return List of matching users
     */
    public List<User> searchUsers(String query) {
        logger.debug("Searching users: " + query);

        String searchPattern = "%" + query + "%";
        return db.query(
            "SELECT * FROM users WHERE email LIKE ? OR name LIKE ?",
            searchPattern, searchPattern
        );
    }

    /**
     * Generate a unique user ID
     *
     * @return User ID
     */
    private String generateId() {
        return "user_" + UUID.randomUUID().toString().replace("-", "");
    }
}

File: test-projects/php-sample/src/UserService.php
<?php

namespace CodeGraph\Sample;

use DateTime;

/**
 * User data model
 */
class User
{
    public function __construct(
        public string $id,
        public string $email,
        public string $name,
        public DateTime $createdAt
    ) {}
}

/**
 * Service for managing user operations
 */
class UserService
{
    private DatabaseConnection $db;
    private Logger $logger;

    /**
     * Create a new UserService
     *
     * @param DatabaseConnection $db Database connection instance
     * @param Logger $logger Logger instance
     */
    public function __construct(DatabaseConnection $db, Logger $logger)
    {
        $this->db = $db;
        $this->logger = $logger;
    }

    /**
     * Create a new user
     *
     * @param string $email User's email address
     * @param string $name User's full name
     * @return User The created user
     * @throws \Exception If email or name is invalid
     */
    public function createUser(string $email, string $name): User
    {
        $this->logger->info("Creating user: {$email}");

        if (empty($email) || !filter_var($email, FILTER_VALIDATE_EMAIL)) {
            throw new \Exception('Invalid email address');
        }

        if (empty($name) || strlen($name) < 2) {
            throw new \Exception('Invalid name');
        }

        $user = new User(
            id: $this->generateId(),
            email: $email,
            name: $name,
            createdAt: new DateTime()
        );

        $this->db->execute(
            'INSERT INTO users (id, email, name, created_at) VALUES (?, ?, ?, ?)',
            [$user->id, $user->email, $user->name, $user->createdAt->format('Y-m-d H:i:s')]
        );

        $this->logger->info("User created: {$user->id}");
        return $user;
    }

    /**
     * Get user by ID
     *
     * @param string $userId User ID
     * @return User|null User if found, null otherwise
     */
    public function getUserById(string $userId): ?User
    {
        $this->logger->debug("Fetching user: {$userId}");

        $result = $this->db->query(
            'SELECT * FROM users WHERE id = ?',
            [$userId]
        );

        if (empty($result['rows'])) {
            return null;
        }

        return $this->mapRowToUser($result['rows'][0]);
    }

    /**
     * Update user information
     *
     * @param string $userId User ID
     * @param string|null $email New email address (optional)
     * @param string|null $name New name (optional)
     * @return User Updated user
     * @throws \Exception If user not found or no fields to update
     */
    public function updateUser(string $userId, ?string $email = null, ?string $name = null): User
    {
        $this->logger->info("Updating user: {$userId}");

        $fields = [];
        $values = [];

        if ($email !== null) {
            $fields[] = 'email = ?';
            $values[] = $email;
        }

        if ($name !== null) {
            $fields[] = 'name = ?';
            $values[] = $name;
        }

        if (empty($fields)) {
            throw new \Exception('No fields to update');
        }

        $values[] = $userId;

        $this->db->execute(
            'UPDATE users SET ' . implode(', ', $fields) . ' WHERE id = ?',
            $values
        );

        $updatedUser = $this->getUserById($userId);
        if ($updatedUser === null) {
            throw new \Exception("User not found: {$userId}");
        }

        return $updatedUser;
    }

    /**
     * Delete a user
     *
     * @param string $userId User ID to delete
     */
    public function deleteUser(string $userId): void
    {
        $this->logger->info("Deleting user: {$userId}");
        $this->db->execute('DELETE FROM users WHERE id = ?', [$userId]);
    }

    /**
     * List all users with pagination
     *
     * @param int $limit Maximum number of users to return
     * @param int $offset Number of users to skip
     * @return array<User> List of users
     */
    public function listUsers(int $limit = 10, int $offset = 0): array
    {
        $this->logger->debug("Listing users: limit={$limit}, offset={$offset}");

        $result = $this->db->query(
            'SELECT * FROM users ORDER BY created_at DESC LIMIT ? OFFSET ?',
            [$limit, $offset]
        );

        return array_map(
            fn($row) => $this->mapRowToUser($row),
            $result['rows']
        );
    }

    /**
     * Search users by email or name
     *
     * @param string $query Search query
     * @return array<User> List of matching users
     */
    public function searchUsers(string $query): array
    {
        $this->logger->debug("Searching users: {$query}");

        $searchPattern = "%{$query}%";
        $result = $this->db->query(
            'SELECT * FROM users WHERE email LIKE ? OR name LIKE ?',
            [$searchPattern, $searchPattern]
        );

        return array_map(
            fn($row) => $this->mapRowToUser($row),
            $result['rows']
        );
    }

    /**
     * Generate a unique user ID
     *
     * @return string User ID
     */
    private function generateId(): string
    {
        $timestamp = (int)(microtime(true) * 1000);
        $random = bin2hex(random_bytes(4));
        return "user_{$timestamp}_{$random}";
    }

    /**
     * Map database row to User object
     *
     * @param array $row Database row
     * @return User User object
     */
    private function mapRowToUser(array $row): User
    {
        return new User(
            id: $row['id'],
            email: $row['email'],
            name: $row['name'],
            createdAt: new DateTime($row['created_at'])
        );
    }
}
